src/AggregateFunctions/AggregateFunctionGroupArraySorted.h

```diff
@@ -0,0 +1,390 @@
+        , max_elems(max_elems_)
+        , seed(seed_)
+    {
+    }
+
+    String getName() const override { return "groupArraySorted"; }
+
+    void create(AggregateDataPtr __restrict place) const override /// NOLINT
+    {
+        [[maybe_unused]] auto a = new (place) Data;
+    }
+
+    void add(AggregateDataPtr __restrict place, const IColumn ** columns, size_t row_num, Arena * arena) const override
+    {
+        const auto & row_values = assert_cast<const ColumnVector<T> &>(*columns[0]).getData();
+
+        if (limit_num_elems && row_values.size() < max_elems)
+            throw Exception(ErrorCodes::INCORRECT_DATA, "The max size of result array is bigger than the actual array size");
+
+        const auto & row_value = row_values[row_num];
+        auto & cur_elems = this->data(place);
+
+        ++cur_elems.total_values;
+        cur_elems.value.push_back(row_value, arena);
+
+        if (!limit_num_elems && columns[0]->size() == cur_elems.value.size())
+            std::sort(cur_elems.value.begin(), cur_elems.value.end());
+        else if (columns[0]->size() == cur_elems.value.size())
+        {
+            /// To optimize, we sort (2 * max_size) elements of input array over and over again
+            /// and after each loop we delete the last half of sorted array
+            bool sorted = false, part_sorted = false;
+            while (!sorted)
+            {
+                if (cur_elems.value.size() >= max_elems * 2)
+                {
+                    std::sort(cur_elems.value.begin(), cur_elems.value.begin() + (max_elems * 2));
+                    cur_elems.value.erase(cur_elems.value.begin() + max_elems, cur_elems.value.begin() + (max_elems * 2));
+                    part_sorted = true;
+                }
+                else if (cur_elems.value.size() > max_elems)
+                {
+                    std::sort(cur_elems.value.begin(), cur_elems.value.end());
+                    cur_elems.value.resize(max_elems, arena);
+                    sorted = true;
+                }
+                else if (cur_elems.value.size() == max_elems)
+                {
+                    if (part_sorted)
+                        sorted = true;
+                    else
+                    {
+                        std::sort(cur_elems.value.begin(), cur_elems.value.end());
+                        sorted = true;
+                    }
+                }
+            }
+        }
+    }
+
+    void merge(AggregateDataPtr __restrict place, ConstAggregateDataPtr rhs, Arena * arena) const override
+    {
+        auto & cur_elems = this->data(place);
+        size_t cur_size = cur_elems.value.size();
+        auto & rhs_elems = this->data(rhs);
+
+        if (rhs_elems.value.empty())
+            return;
+
+        if (limit_num_elems)
+        {
+            UInt64 elems_to_insert = std::min(static_cast<size_t>(max_elems) - cur_size, rhs_elems.value.size());
```

Comment:
I do not understand why this is needed.
You need to get top `max_elems` smallest numbers from 2 (unsorted) arrays (of size <= max_elems * 2).
(current array is mutable, rhs is constant).

The easiest way is to insert all the values from rhs to lhs, sort lhs, and remove no needed values.
Maybe we can do it in a few steps (insert the first max_elems, sort/remove, insert the second max_elems, sort/remove) to keep memory usage of state <= `2 * max_elements * sizeof(T)`.